db/clientlist_virtual_reader.cc

```diff
@@ -0,0 +1,237 @@
+                    cluster.begin(), cluster.end(),
+                    [] (const auto& client1, const auto& client2) { return client1.port > client2.port; });
+        } else {
+            std::sort(
+                    cluster.begin(), cluster.end(),
+                    [] (const auto& client1, const auto& client2) { return client1.port < client2.port; });
+        }
+        sorted_clients.insert(
+                sorted_clients.end(),
+                make_move_iterator(cluster.begin()),
+                make_move_iterator(cluster.end()));
+    }
+    return make_ready_future<std::vector<system_keyspace::client_data>>(std::move(sorted_clients));
+}
+
+future<> clientlist_mutation_reader::maybe_setup_mutation_fragmentator() {
+    if (_mutation_fragmentator_opt) {
+        return make_ready_future<>();
+    }
+
+    return smp::submit_to(0, [] {
+        // It must be run on shard 0. Shards with id>=1 see: get_dist_cql_server()==nullptr
+        return std::ref(*service::get_local_storage_service().get_dist_cql_server());
+    }).then([] (distributed<cql_transport::cql_server>& distributed_cql_server) {
+        return get_all_clients_data(distributed_cql_server);
+    }).then([this] (auto&& client_data_vec) {
+        return filter_and_sort_by_port(std::move(client_data_vec)).then([this] (auto&& sorted_client_data_vec) {
+            return do_with(
+                    std::vector<mutation>{},
+                    [this, scdv=std::move(sorted_client_data_vec)] (auto&& mutations_vec) mutable {
+                mutations_vec.reserve(scdv.size());
+
+                std::for_each(scdv.begin(), scdv.end(), [this, &mutations_vec] (auto&& client) {
+                    mutations_vec.emplace_back(db::system_keyspace::make_client_mutation(std::move(client)));
+                });
+                if (!mutations_vec.empty()) {
+                    _mutation_fragmentator_opt = flat_mutation_reader_from_mutations(std::move(mutations_vec));
+                }
+                return make_ready_future<>();
+            });
+        });
+    });
+}
+
+clientlist_mutation_reader::clientlist_mutation_reader(
+        schema_ptr schema,
+        const dht::partition_range& prange,
+        const query::partition_slice& slice)
+    : impl(schema)
+    , _schema(std::move(schema))
+    , _prange(&prange)
+    , _slice(slice) {}
+
+future<> clientlist_mutation_reader::fill_buffer(db::timeout_clock::time_point timeout) {
+    // PKs returned from different shards must be distinct, otherwise combining reader
+    // hangs up. Now: one client can maintain concurrent connections with many different
+    // shards but his IP belongs to partition range of a particular (possibly yet another)
+    // shard. That's why every shard must traverse every connection and return only the
+    // clients with IPs that fall within *this shard's* partition range.
+
+    return with_timeout(timeout, maybe_setup_mutation_fragmentator()).then([this, timeout] {
```

Comment:
`with_timeout()` seems wrong, what prevents `clientlist_mutation_reader` from being destroyed after timeout while `maybe_setup_mutation_fragmentator()` is still executing (which results in use-after-free)?